NEW TEST(276140@main): [ MacOS WK1 Debug ] media/media-source/worker/media-managedmse-worker.html  is a constant crash

jyavenard opened the pull request:
#### b7d8be7a8274318250bdbb161a6eee6daab33991
<pre>
NEW TEST(276140@main): [ MacOS WK1 Debug ] media/media-source/worker/media-managedmse-worker.html  is a constant crash
<a href="https://bugs.webkit.org/show_bug.cgi?id=271323">https://bugs.webkit.org/show_bug.cgi?id=271323</a>
<a href="https://rdar.apple.com/125100787">rdar://125100787</a>

Reviewed by Youenn Fablet.

WebPreferences are always set when running LayoutTests. MediaSource in a Worker is not functional
unless the GPU process is active and the MediaPlayers are to run in the GPU process.
MediaSource in a Worker can only work with a MediaSourcePrivate that is designed to be thread-safe
and at present there&apos;s only one kind: MediaSourcePrivateRemote.
As it is possible for a WebPreference to be turned on, we need to handle the case where it has
been accidentally set on a non-supported platform.
To achieve this we add a new MediaStrategy API: `hasThreadSafeSupport()` which will be checked
in addition to the `mediaSourceInWorkerEnabled` preference.

Updated existing test to ensure that enabling the pref on a non-supported configuration is
behaving properly.

* LayoutTests/media/media-source/worker/media-managedmse-worker-expected.txt:
* LayoutTests/media/media-source/worker/media-managedmse-worker.html:
* LayoutTests/media/media-source/worker/worker.js:
(onmessage):
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk1/media/media-source/worker/media-managedmse-worker-expected.txt: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::enabledForContext):
(WebCore::MediaSource::canConstructInDedicatedWorker):
* Source/WebCore/platform/MediaStrategy.cpp:
(WebCore::MediaStrategy::hasThreadSafeMediaSourceSupport const):
* Source/WebCore/platform/MediaStrategy.h:
* Source/WebCore/platform/PlatformStrategies.h:
(WebCore::PlatformStrategies::mediaStrategy): Allow creation of the MediaStrategy object on any threads, in practice this MediaStrategy creation is only
ever done on the main thread, but this could be different in the future.
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
(WebKit::WebMediaStrategy::createAudioDestination):
(WebKit::WebMediaStrategy::createNowPlayingManager const):
(WebKit::WebMediaStrategy::hasThreadSafeMediaSourceSupport const):
(WebKit::WebMediaStrategy::enableMockMediaSource):
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.h:

Canonical link: <a href="https://commits.webkit.org/276534@main">https://commits.webkit.org/276534@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/59819db0b9986d3ffb56abbc54f3a9d09670da28

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44882 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23981 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47372 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47536 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40887 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47185 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28037 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21378 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36849 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45460 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21031 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/38643 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17935 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18443 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39787 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2929 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/38082 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41111 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40080 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49206 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/44338 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19850 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16395 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43843 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21171 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/38880 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42605 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9997 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/21507 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/51510 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20843 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10453 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/MediaStrategy.cpp:
```diff
@@ -51,6 +51,11 @@ void MediaStrategy::resetMediaEngines()
     m_mockMediaSourceEnabled = false;
 }
 
+bool MediaStrategy::hasThreadSafeMediaSupport() const
```
```suggestion
bool MediaStrategy::hasThreadSafeMediaSourceSupport() const
```

jyavenard replied:
It's the MediaPlayerPrivateRemote itself that needs to be thread-safe, not just MSE code. but seeing indeed it's only used in combination with MSE, I'll consider that name

youennf commented on Source/WebCore/platform/MediaStrategy.cpp:
```diff
@@ -72,6 +77,7 @@ void MediaStrategy::addMockMediaSourceEngine()
 {
     MediaPlayerFactorySupport::callRegisterMediaEngine(MockMediaPlayerMediaSource::registerMediaEngine);
 }
+
```
Unneeded

youennf commented on Source/WebCore/platform/graphics/MediaPlayer.cpp:
```diff
@@ -1392,6 +1396,16 @@ bool MediaPlayer::supportsKeySystem(const String& keySystem, const String& mimeT
+{
+#if USE(AVFOUNDATION)
+    Locker locker { s_lock };
+    return !!registerRemotePlayerCallback();
```
It is a bit weird to call the callback in a bool getter.

jyavenard replied:
At this stage, it is the only data available in WebCore that identify media in the GPU process being used. And it is the presence of that callback that is used throughout MediaPlayer and other code to determine if we are to use the GPU process.

so a bit weird yes, but existing, I only added the existence of the callback.
The presence of the callback is the ultimate source of truth to determine if the GPU process will be used for playback

jyavenard replied:
ok, the callback is set if the pref `UseGPUProcessForMediaEnabled` is set, and where the callback is set is done just after setting the m_useGPUProcess on the MediaStrategy, so it's equivalent; I've removed the code entirely.

youennf commented on Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
```diff
@@ -88,9 +90,19 @@ std::unique_ptr<WebCore::NowPlayingManager> WebMediaStrategy::createNowPlayingMa
+bool WebMediaStrategy::hasThreadSafeMediaSupport() const
+{
+#if ENABLE(GPU_PROCESS)
+    return m_useGPUProcess && WebCore::MediaPlayer::hasThreadSafeSupport();
```
Why isn't m_useGPUProcess sufficient? I would guess that if m_useGPUProcess is true, we are using remote media players.

jyavenard replied:
because you can have the GPU process and not use it for media. Even on cocoa, this is a pref `UseGPUProcessForMediaEnabled` while it's true by default, the entire reason for this bug is due to pref having unexpected values. So we have to cater for the case where UseGPUProcessForMediaEnabled was disabled. It's still doable in the UI.

Until that pref is removed, we need to test for both m_useGPUProcess and if it's active in the MediaPlayer.

I guess I could use UseGPUProcessForMediaEnabled instead ; but the pref doesn't exist for GTK so it will require some #ifdef-fu it's also a WebKit only pref.

youennf reviewed (approved):
LGTM.
I wonder whether overriding mediaSourceInWorkerEnabled setting to false if GPU process is off would be simpler or not.

jyavenard committed:
NEW TEST(276140@main): [ MacOS WK1 Debug ] media/media-source/worker/media-managedmse-worker.html  is a constant crash
https://bugs.webkit.org/show_bug.cgi?id=271323
rdar://125100787

Reviewed by Youenn Fablet.

WebPreferences are always set when running LayoutTests. MediaSource in a Worker is not functional
unless the GPU process is active and the MediaPlayers are to run in the GPU process.
MediaSource in a Worker can only work with a MediaSourcePrivate that is designed to be thread-safe
and at present there's only one kind: MediaSourcePrivateRemote.
As it is possible for a WebPreference to be turned on, we need to handle the case where it has
been accidentally set on a non-supported platform.
To achieve this we add a new MediaStrategy API: `hasThreadSafeSupport()` which will be checked
in addition to the `mediaSourceInWorkerEnabled` preference.

Updated existing test to ensure that enabling the pref on a non-supported configuration is
behaving properly.

* LayoutTests/media/media-source/worker/media-managedmse-worker-expected.txt:
* LayoutTests/media/media-source/worker/media-managedmse-worker.html:
* LayoutTests/media/media-source/worker/worker.js:
(onmessage):
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk1/media/media-source/worker/media-managedmse-worker-expected.txt: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::enabledForContext):
(WebCore::MediaSource::canConstructInDedicatedWorker):
* Source/WebCore/platform/MediaStrategy.cpp:
(WebCore::MediaStrategy::hasThreadSafeMediaSourceSupport const):
* Source/WebCore/platform/MediaStrategy.h:
* Source/WebCore/platform/PlatformStrategies.h:
(WebCore::PlatformStrategies::mediaStrategy): Allow creation of the MediaStrategy object on any threads, in practice this MediaStrategy creation is only
ever done on the main thread, but this could be different in the future.
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.cpp:
(WebKit::WebMediaStrategy::createAudioDestination):
(WebKit::WebMediaStrategy::createNowPlayingManager const):
(WebKit::WebMediaStrategy::hasThreadSafeMediaSourceSupport const):
(WebKit::WebMediaStrategy::enableMockMediaSource):
* Source/WebKit/WebProcess/GPU/media/WebMediaStrategy.h:

Canonical link: https://commits.webkit.org/276534@main
